src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicates.java

```diff
@@ -0,0 +1,203 @@
+import org.broadinstitute.hellbender.cmdline.Argument;
+import org.broadinstitute.hellbender.cmdline.CommandLineProgramProperties;
+import org.broadinstitute.hellbender.cmdline.programgroups.SparkProgramGroup;
+import org.broadinstitute.hellbender.cmdline.programgroups.TestSparkProgramGroup;
+import org.broadinstitute.hellbender.engine.spark.GATKSparkTool;
+import org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadCoordinateComparator;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+import scala.Tuple2;
+
+import java.io.Serializable;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@CommandLineProgramProperties(summary = "Compares two BAMs for duplicates", oneLineSummary = "Compares two BAMs for duplicates",
+        programGroup = TestSparkProgramGroup.class)
+
+public final class CompareDuplicates extends GATKSparkTool {
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public boolean requiresReads() { return true; }
+
+    @Argument(doc="the second BAM", shortName = "I2", fullName = "input2", optional = false)
+    protected String input2;
+
+    @Argument(doc="print summary", shortName = "ps", fullName = "printSummary", optional = true)
+    protected Boolean printSummary = true;
+
+    @Argument(doc="crash on diff", shortName = "cd", fullName = "crashOnDiff", optional = true)
+    protected Boolean crashOnDiff = false;
+
+    @Override
+    protected void runTool(final JavaSparkContext ctx) {
+        // We only need to compare duplicates that are "primary" (i.g., primary mapped read).
+        JavaRDD<GATKRead> firstReads = getReads().filter(v1 -> {
+            if (Utils.isNonPrimary(v1) && v1.isDuplicate()) {
+                throw new GATKException("found a non-primary read marked as a duplicate in the first bam (this shouldn't happen)");
+            }
+            return !Utils.isNonPrimary(v1);
+        });
+        firstReads.foreach(GATKRead::clearAttributes);
+
+        ReadsSparkSource readsSource2 = new ReadsSparkSource(ctx);
+        JavaRDD<GATKRead> secondReads = readsSource2.getParallelReads(input2, getIntervals()).filter(v1 -> {
+            if (Utils.isNonPrimary(v1) && v1.isDuplicate()) {
+                throw new GATKException("found a non-primary read marked as a duplicate in the second bam (this shouldn't happen)");
+            }
+            return !Utils.isNonPrimary(v1);
+        });
+        secondReads.foreach(GATKRead::clearAttributes);
+
+
+        // Start by verifying that we have same number of reads and duplicates in each BAM.
+        long firstBamSize = firstReads.count();
+        JavaRDD<GATKRead> firstDupes = firstReads.filter(GATKRead::isDuplicate);
```

Comment:
inline this variable (and secondDupes) - they add nothing and make me think they are used downstream, which they are not